paper.tex

```diff
@@ -449,14 +455,83 @@ \subsection{Pipeline Design} \label{subsec:pipeline-design}
+            }
+            \If{cusp0 < 0 OR t < 1}{
+                evolute\_start(es, offset)\;
+                range $\gets$ cusp0 >= 0 ? vec2(t, 1) : vec2(0, t)\;
+                es\_seg\_flatten\_evolute(es, range)\;
+                es\_seg\_flatten\_offset\_reverse(es, offset, range)\;
+            }
+            \If{cusp0 < 0 AND t < 1}{
+                evolute\_finalize()\;
+                es\_seg\_flatten\_offset(es, offset, vec2(t, 1))\;
+            }
+            \tcp{Pop the stack}
+            t0\_u $\gets$ t0\_u + 1u\;
+            shift $\gets$ countTrailingZeros(t0\_u)\;
+            t0\_u $\gets$ t0\_u >> shift;
+            dt $\gets$ dt * f32(1u << shift)\;
+            last\_p $\gets$ next\_p\;
+            last\_q $\gets$ next\_q\;
+        }{
+            \tcp{Error is above the tolerance.}
+            \tcp{Push the stack; Continue subdivision}
+            t0\_u $\gets$ t0\_u * 2u\;
+            dt $\gets$ dt * 0.5\;
+        }
+    }
+}
+\caption{Outline of the iterative cubic Bézier flattening routine invoked in our compute shader. A working WGSL implementation is provided in the supplemental materials.}
+\label{alg:flatten-cubic}
+\end{algorithm}
+
+All input segments first get converted to a cubic Bézier by degree elevation. Our flattening routine (see Algorithm~\ref{alg:flatten-cubic}) iteratively subdivides the input cubic segment in half, and at each step, computes an error estimate for geometric Hermite interpolation to an Euler spiral segment. Once the error satisfies a desired tolerance threshold, the Euler spiral segment is directly flattened to a polyline (or, alternately, converted to a sequence of arc segments) by invoking a subroutine that is parameterized by the stroke halfwidth (see \texttt{es\_seg\_flatten\_offset} in Algoritm~\ref{alg:flatten-cubic}). For a regular fill, this parameter is set to $0$. For strokes, the sign of the parameter determines the position of the parallel curve relative to the source segment as well as the winding direction of the output lines.
 
-The joins between adjacent segments need to form a continuous outline when combined with the segments. This requires that a thread have access to the tangent vector at the end of its assigned segment \emph{and} the start tangent of the next adjacent segment. We require that the input segments within a subpath are stored in order which simplifies the access to the adjacent segment down to a buffer read at the next array offset. While we take care to output the \emph{outer} join segment (where the offset curves meet at a reflex angle) according to the join style, we output the inner join simply as a line segment.
+This procedure is sufficient to generate weakly correct stroke outlines. If the outline contains a cusp -- which can be determined by evaluating the curvature variation along the Euler spiral segment -- we output an evolute patch (as shown in Figure~\ref{fig:evolutes}) in order to achieve strong correctness.
```

Comment:
> -- which can be determined by evaluating the curvature variation along the Euler spiral segment --

There might be a better way to summarize the computation of `t` (the evolute anchor point / cusp) via the `cusp0 * cusp1` multiplication and the sign comparison.  I think the new algorithm block demonstrates it clearly enough but maybe this should explain the state machine too.